library/seq.lisp

```diff
@@ -0,0 +1,517 @@
+(coalton-library/utils:defstdlib-package #:coalton-library/seq
+  (:use
+   #:coalton
+   #:coalton-library/builtin
+   #:coalton-library/functions
+   #:coalton-library/classes)
+  (:local-nicknames
+   (#:types #:coalton-library/types)
+   (#:math #:coalton-library/math/integral)
+   (#:optional #:coalton-library/optional)
+   (#:cell #:coalton-library/cell)
+   (#:vector #:coalton-library/vector)
+   (#:iter #:coalton-library/iterator))
+  (:export
+   #:Seq
+   #:new
+   #:push
+   #:pop
+   #:size
+   #:get
+   #:put
+   #:empty?
+   #:conc
+   #:make))
+
+(in-package #:coalton-library/seq)
+
+(named-readtables:in-readtable coalton:coalton)
+
+#+coalton-release
+(cl:declaim #.coalton-impl/settings:*coalton-optimize-library*)
+
+;;;;
+;;; This module implements a persistent sequence data type based on
+;;; Relaxed Radix Balanced Trees, introduced by Phillip Bagwell
+;;; See: 
+;;; https://www.semanticscholar.org/paper/RRB-Trees%3A-Efficient-Immutable-Vectors-Bagwell-Rompf/30c8c562f6421ab6b00d0b7faebd897c407de69c
+;;; 
+
+;;; Private utility macro for convenient matching of a single case of a ADT
+(cl:eval-when (:compile-toplevel :load-toplevel :execute)
+  (cl:defmacro match1 (pattern form cl:&body body)
+    `(match ,form
+       (,pattern ,@body)
+       (_ (unreachable)))))
+
+(coalton-toplevel 
+  ;;
+  ;; Seq
+  ;; 
+  
+  (declare max-branching UFix)
+  (define max-branching 32)
+
+  (declare min-branching UFix)
+  (define min-branching 24)
+
+  (define-type (Seq :a)
+    (RelaxedNode
+     UFix                               ; height
+     UFix                               ; cached full subtree size
+     (vector:Vector UFix)               ; cumulative size table
+     (vector:Vector (Seq :a)))         ; subtrees
+    (LeafArray (vector:Vector :a)))
+
+  (declare new (types:RuntimeRepr :a => Unit -> Seq :a))
+  (define (new)
+    "Create a new empty Seq."
+    (LeafArray (vector:new)))
+
+  (declare size (Seq :a -> UFix))
+  (define (size seq)
+    "Return the number of elements in the sequence."
+    (match seq
+      ((RelaxedNode _ _ cst _) (vector:last-unsafe cst))
+      ((LeafArray leaves) (vector:length leaves))))
+  
+  (define (empty? seq)
+    (== 0 (size seq)))
+
+  (declare get (Seq :a -> UFix -> Optional :a))
+  (define (get seq idx)
+    "Get the member of SEQ at index IDX, or None if IDX is larger
```

Comment:
Coalton docstring style is to use markdown syntax and lowercase names

    Get the member of `seq` at ...